### PR TITLE
Jd/fixed output

### PR DESCRIPTION
**Pull request by jd-lara**
I am making this PR now because it is getting really large and it still doesn't address the requirements but it standardizes more the fixed output. 

Also, it fixes several issues of unimplemented code features.

### CODE REVIEW

File: src/services_models/reserves.jl [hunk 1]
```diff
@@ -99,6 +91,17 @@ function modify_device_model!(
...
 
+function include_service!(
+    constraint_data::DeviceTimeSeries,
+    services::Vector{<:PSY.Reserve},
+    SM::ServiceModel,
+)
+    range_data = constraint_data.range
+    isnothing(range_data) && return
+    include_service!(range_data, services, SM)
+    return
```
Comment (daniel-thom) [inline src/services_models/reserves.jl:102]: nit: IMHO fewer returns is better

```
if !isnothing(range_data)
    include_service!(range_data, services, SM)
end
return
```

File: src/services_models/reserves.jl [hunk 2]
```diff
@@ -143,3 +145,35 @@ function _device_services!(
...
+function add_device_services!(
+    constraint_data_in::RangeConstraintsData,
+    constraint_data_out::RangeConstraintsData,
+    device::D,
+    model::DeviceModel{D, <:AbstractStorageFormulation},
+) where {D <: PSY.Storage}
+    for service_model in get_services(model)
+        if PSY.has_service(device, service_model.service_type)
+            services =
+                [s for s in PSY.get_services(device) if isa(s, service_model.service_type)]
```
Comment (daniel-thom) [inline src/services_models/reserves.jl:158]: You don't actually need a vector. You could use a generator instead.

Minor readability thing:  I find `s isa service_model.service_type` more readable.  Just my opinion.

File: src/services_models/reserves.jl [hunk 3]
```diff
@@ -143,3 +145,35 @@ function _device_services!(
...
+            @assert !isempty(services)
+            if service_model.service_type <: PSY.Reserve{PSY.ReserveDown}
+                for (ix, service) in enumerate(services)
+                    push!(
+                        constraint_data_in.additional_terms_ub,
+                        constraint_name(PSY.get_name(service), service_model.service_type),
+                    )
+                end
+            end
+            if service_model.service_type <: PSY.Reserve{PSY.ReserveUp}
```
Comment (daniel-thom) [inline src/services_models/reserves.jl:168]: elseif

File: src/services_models/reserves.jl [hunk 4]
```diff
@@ -143,3 +145,35 @@ function _device_services!(
...
+    device::D,
+    model::DeviceModel{D, <:AbstractStorageFormulation},
+) where {D <: PSY.Storage}
+    for service_model in get_services(model)
+        if PSY.has_service(device, service_model.service_type)
+            services =
+                [s for s in PSY.get_services(device) if isa(s, service_model.service_type)]
+            @assert !isempty(services)
+            if service_model.service_type <: PSY.Reserve{PSY.ReserveDown}
+                for (ix, service) in enumerate(services)
```
Comment (daniel-thom) [inline src/services_models/reserves.jl:161]: You aren't using `ix` here or in the next clause.  `for service in services`

File: src/devices_models/device_constructors/renewablegeneration_constructor.jl [hunk 5]
```diff
@@ -63,17 +61,16 @@ end
 function construct_device!(
     psi_container::PSIContainer,
     sys::PSY.System,
-    model::DeviceModel{R, RenewableFixed},
-    system_formulation::Type{S};
-    kwargs...,
+    model::DeviceModel{R, FixedOutput},
+    ::Type{S},
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/devices_models/device_constructors/renewablegeneration_constructor.jl:65]: This is sort of at odds with the idea that the first argument should be the one mutated. E.g. `push!()`

File: src/devices_models/devices/common/add_parameters.jl [hunk 6]
```diff
@@ -1,34 +1,12 @@
...
     psi_container::PSIContainer,
     ts_data::Vector{DeviceTimeSeries},
     param_reference::UpdateRef,
     expression_name::Symbol,
     multiplier::Float64 = 1.0,
 )
-    if !model_has_parameters(psi_container)
-        throw(IS.DataFormatError("Operational Model doesn't have parameters enabled. Include the keyword use_parameters=true"))
-    end
+    @assert model_has_parameters(psi_container)
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/devices_models/devices/common/add_parameters.jl:9]: No, this is an internal function that is meant for developers to call. It has helped to catch bad uses of `add_parameters` when the implementation is incorrect.

File: src/utils.jl [hunk 7]
```diff
@@ -271,3 +271,7 @@ end
 function remove_chars(s::String, char::String)
     return replace_chars(s::String, char::String, "")
 end
+
+function get_available_components(::Type{T}, sys::PSY.System) where {T <: PSY.Component}
+    return PSY.get_components(T, sys, x -> PSY.get_available(x))
+end
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/utils.jl:277]: I will leave the function until we merge https://github.com/NREL-SIIP/PowerSystems.jl/pull/459

File: src/services_models/reserves.jl (hunk 2 above)
Comment (jd-lara) [reply src/services_models/reserves.jl:158]: we use isa() as a function throughout the code in PSI.

File: src/devices_models/device_constructors/renewablegeneration_constructor.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (daniel-thom) [reply src/devices_models/device_constructors/renewablegeneration_constructor.jl:65]: I'm slightly wrong.  https://docs.julialang.org/en/v1/manual/style-guide/index.html#Write-functions-with-argument-ordering-similar-to-Julia-Base-1

File: src/core/psi_container.jl [hunk 8]
```diff
@@ -163,7 +163,7 @@ function psi_container_init!(
     end
 
     if get_use_forecast_data(settings)
-        total_number_of_devices = length(PSY.get_components(PSY.Device, sys))
+        total_number_of_devices = length(get_available_components(PSY.Device, sys))
```
Comment (daniel-thom) [inline src/core/psi_container.jl:166]: Seems like you never want to do anything with components that are not available.  Should we have a flag in PowerSystems that prevents those components from being added?  Maybe this is not appropriate...just an idea to simplify the logic.

File: src/devices_models/devices/hydro_generation.jl [hunk 9]
```diff
@@ -220,10 +174,16 @@ function activepower_constraints!(
     parameters = model_has_parameters(psi_container)
     use_forecast_data = model_uses_forecasts(psi_container)
 
-    ts_data_active, constraint_data =
-        _get_time_series(psi_container, devices, model, x -> PSY.get_activepowerlimits(x))
-
     if !parameters && !use_forecast_data
```
Comment (daniel-thom) [inline src/devices_models/devices/hydro_generation.jl:177]: Is it possible for parameters to be true and use_forecast_data to be false?  This code doesn't handle that case.

File: src/devices_models/devices/renewable_generation.jl [hunk 10]
```diff
@@ -150,14 +94,16 @@ function activepower_constraints!(
...
     use_forecast_data = model_uses_forecasts(psi_container)
 
-    ts_data_active, _, constraint_data = _get_time_series(
-        psi_container,
-        devices,
-        model,
-        x -> (min = 0.0, max = PSY.get_activepower(x)),
-    )
-
     if !parameters && !use_forecast_data
```
Comment (daniel-thom) [inline src/devices_models/devices/renewable_generation.jl:97]: Same comment here

File: src/devices_models/devices/hydro_generation.jl (hunk 9 above)
Comment (jd-lara) [reply src/devices_models/devices/hydro_generation.jl:177]: I don't think that case was ever handled.

File: src/core/psi_container.jl (hunk 8 above)
Comment (jd-lara) [reply src/core/psi_container.jl:166]: I rather leave it as it is and have the users make modifications to their systems if they want to make a device unavailable explicitly.

**Pull request merged**